libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -128,238 +169,771 @@ BeaconSetupService::State BeaconSetupService::OnIdle()
 BeaconSetupService::State BeaconSetupService::OnWaitForDirectConnections()
 {
   std::lock_guard<std::mutex> lock(mutex_);
-  auto                        v_peers = endpoint_.GetDirectlyConnectedPeers();
-  std::unordered_set<Address> peers(v_peers.begin(), v_peers.end());
+  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_DIRECT_CONNECTIONS));
 
-  bool     all_connected = true;
-  uint16_t connected     = 0;
+  auto                              v_peers = endpoint_.GetDirectlyConnectedPeers();
+  std::unordered_set<MuddleAddress> peers(v_peers.begin(), v_peers.end());
+
+  std::set<MuddleAddress> connected;
 
   for (auto &m : beacon_->aeon.members)
   {
     // Skipping own address
     if (m == identity_)
     {
+      connected.insert(m.identifier());
       continue;
     }
 
     // Checking if other peers are there
     if (peers.find(m.identifier()) == peers.end())
     {
-      all_connected = false;
-
       // TODO(tfr): Request muddle to connect.
     }
     else
     {
-      connected++;
+      connected.insert(m.identifier());
     }
   }
 
-  if (all_connected)
+  if (connected.size() == beacon_->aeon.members.size())
   {
-    FETCH_LOG_INFO(LOGGING_NAME, "All peers connected. Proceeding.");
-    return State::BROADCAST_ID;
+    connections_ = connected;
+    fetch::serializers::MsgPackSerializer serializer;
+    serializer << connected;
+    pre_dkg_rbc_.Broadcast(serializer.data());
+    return State::WAIT_FOR_READY_CONNECTIONS;
   }
 
   state_machine_->Delay(std::chrono::milliseconds(200));
-  FETCH_LOG_INFO(LOGGING_NAME,
-                 "Waiting for all peers to join before starting setup. Connected: ", connected,
-                 " expect: ", beacon_->aeon.members.size() - 1);
+  FETCH_LOG_INFO(LOGGING_NAME, "Waiting for all peers to join before starting setup. Connected: ",
+                 connected.size(), " expect: ", beacon_->aeon.members.size() - 1);
 
   return State::WAIT_FOR_DIRECT_CONNECTIONS;
 }
 
-BeaconSetupService::State BeaconSetupService::OnBroadcastID()
+BeaconSetupService::State BeaconSetupService::OnWaitForReadyConnections()
 {
   std::lock_guard<std::mutex> lock(mutex_);
+  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_READY_CONNECTIONS));
 
-  // Creating identity
-  CabinetMemberDetails member;
-  member.identity = beacon_->manager.identity();
-  member.id       = beacon_->manager.id();
-  member_details_queue_.push_back(member);
+  // Try broadcasting connections again in case they were discarded by RBCs who's cabinet was reset
```

Comment:
This might deadlock the network in the limit right? Is there no feedback mechanism from the DKG to indicate how many peers have recieved your message?

Reply:
You are correct. This is not the way to solve this but we need some feedback mechanism between the RBC and DKG for this. Or we make the nodes send requests to peers which it has not received ready messages from.